[MLIR][LLVM] Add ftz and fuse FP ops related function attribute support

runseny opened the pull request:
Adds  `denormal-fp-math-f32`, `denormal-fp-math`, `fp-contract` to llvmFuncOp attributes.

`denormal-fp-math-f32` and `denormal-fp-math` can enable the ftz, that is , flushing denormal to zero.

`fp-contract` can enable the fma fusion such as `mul + add -> fma`

runseny committed:
[MLIR][LLVM] Add ftz and fuse FP ops related function attribute support

runseny commented:
@asb @zero9178 Could you help review this mr? This mr is similar with the previous work: https://github.com/llvm/llvm-project/pull/79812, Thanks in advance!

gysit commented on mlir/test/Target/LLVMIR/Import/function-attributes.ll:
```diff
@@ -344,6 +344,72 @@ declare void @func_attr_no_signed_zeros_fp_math_true() "no-signed-zeros-fp-math"
+
+; CHECK-LABEL: @func_attr_denormal_fp_math_ieee
+; CHECK-SAME: attributes {denormal_fp_math = "ieee"}
+declare void @func_attr_denormal_fp_math_ieee() "denormal-fp-math"="ieee"
```
Since these are string attributes there is probably not much value in testing different string values for every attribute (i.e. ieee, preserve-sign)? Or is there specific reason for testing the different strings? Instead I would have a test for every attribute with possible different strings? So three tests here and three tests in the file below which exercises the lowering to LLVM IR.

runseny replied:
> Or is there specific reason for testing the different strings?

These different strings are the general cmd options in codegen such as [denormalFpMath related](https://github.com/llvm/llvm-project/blob/main/llvm/lib/CodeGen/CommandFlags.cpp#L259) and [fp-contract](https://github.com/llvm/llvm-project/blob/main/llvm/lib/CodeGen/CommandFlags.cpp#L301). So i tested all these general cmd options for functional completeness.

gysit replied:
Ok! I saw there is already prior art above. 

I think the point I tried to make is that testing different values would make a lot of sense if the stored value would be an enum. Then we want to test that all enum values exist and are handled properly. In this case, we store a string so if one works all other strings should work as well (there are no separate code paths for different enum values). For maintenance reasons, I would thus keep the test set minimal if there is no extra functionality (in the sense of code coverage) that is tested.

gysit commented on mlir/lib/Target/LLVMIR/ModuleImport.cpp:
```diff
@@ -1685,6 +1685,9 @@ static constexpr std::array kExplicitAttributes{
     StringLiteral("vscale_range"),
+    StringLiteral("denormal-fp-math"),
+    StringLiteral("denormal-fp-math-f32"),
+    StringLiteral("fp-contract"),
```
nit: Can you sort the array alphabetically. It looks like it was sorted before.

runseny replied:
Yeah, done

gysit commented on mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:
```diff
@@ -1351,6 +1351,18 @@ LogicalResult ModuleTranslation::convertOneFunction(LLVMFuncOp func) {
     llvmFunc->addFnAttr("no-signed-zeros-fp-math",
                         llvm::toStringRef(*noSignedZerosFpMath));
 
+  if (auto DenormalFPMath = func.getDenormalFpMath())
```
```suggestion
  if (auto denormalFPMath = func.getDenormalFpMath())
```
nit: the variable names should start lowercase (below as well).

runseny replied:
Thanks, modified. The diff please see: [diff](https://github.com/llvm/llvm-project/compare/94bd71f97c6e35fac774611554d0069cc0e16836..c296c22b72f24c37acf37614b3f5001f1eb6475d#diff-0e9fd6478c7a83100254a3b40e2d97649a8d41c84506e1da7f8d281fc633b1f5R1354)

gysit commented on mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td:
```diff
@@ -1456,7 +1456,10 @@ def LLVM_LLVMFuncOp : LLVM_Op<"func", [
+    OptionalAttr<UnitAttr>:$optimize_none,
+    OptionalAttr<StrAttr>:$denormal_fp_math,
+    OptionalAttr<StrAttr>:$denormal_fp_math_f32,
+    OptionalAttr<StrAttr>:$fp_contract
```
nit: Are these flags related to the other fp_math flags above? Would it make sense to group them together?

runseny replied:
Yeah, these flags are related to fp_math,  so i grouped these flags with other fastMath flags.

gysit reviewed (commented):
Thanks!

I added some mostly nit comments.

gysit reviewed (approved):
LGTM

Thanks for addressing the comments! I still think there is some redundancy in terms of test cases. Feel free to trim them down if you think my argument makes sense, or otherwise keep them :).

runseny commented:
> LGTM
> 
> Thanks for addressing the comments! I still think there is some redundancy in terms of test cases. Feel free to trim them down if you think my argument makes sense, or otherwise keep them :).

Yeah, you are right, thanks for this suggestion. Now i only test one string for each attribute.

runseny commented:
@zero9178 Hi, i still need an extra approval, Could you help take a look?

zero9178 reviewed (approved):
Not super familiar with these fp functions, but given it mirrors LLVM, its their mess to figure and first and foremost :slightly_smiling_face: 
LGTM with clang-format addressed. Feel free to ping if no one merged this in a day once CI is green.

Given that some of these only have a finite set of limited values afaik, we could turn these into an enum attribute in the future. But this can be a future iteration.

> @zero9178 Hi, i still need an extra approval, Could you help take a look?

FYI, there is no such general rule in LLVM that you need to have more than one approval, although you're of course free to get more eyes on the review :slightly_smiling_face:! Sometimes the first reviewer will ask you for either more time before landing or another approval if they do not feel comfortable being the only approval, but that is not applicable here.

runseny commented:
> Not super familiar with these fp functions, but given it mirrors LLVM, its their mess to figure and first and foremost 🙂 LGTM with clang-format addressed. Feel free to ping if no one merged this in a day once CI is green.
> 
> Given that some of these only have a finite set of limited values afaik, we could turn these into an enum attribute in the future. But this can be a future iteration.
> 
> > @zero9178 Hi, i still need an extra approval, Could you help take a look?
> 
> FYI, there is no such general rule in LLVM that you need to have more than one approval, although you're of course free to get more eyes on the review 🙂! Sometimes the first reviewer will ask you for either more time before landing or another approval if they do not feel comfortable being the only approval, but that is not applicable here.

Thanks for your detail explanation. Now, the CI of this mr passed, could you help merge it? thanks. @zero9178
